src/com/xilinx/rapidwright/router/VersalClockRouting.java

```diff
@@ -0,0 +1,494 @@
+            startingPoints.computeIfAbsent(cr, k -> new HashSet<>())
+                    .add(new NodeWithPrevAndCost(distLine));
+        }
+        return startingPoints;
+    }
+
+    public static void routeToLCBs(Net clk, Map<ClockRegion, Set<NodeWithPrevAndCost>> startingPoints, Set<Node> lcbTargets) {
+        Queue<NodeWithPrevAndCost> q = new PriorityQueue<>();
+        Set<PIP> allPIPs = new HashSet<>();
+        Set<Node> visited = new HashSet<>();
+
+        nextLCB: for (Node lcb : lcbTargets) {
+            q.clear();
+            visited.clear();
+            Tile lcbTile = lcb.getTile();
+            ClockRegion currCR = lcbTile.getClockRegion();
+            Set<NodeWithPrevAndCost> starts = startingPoints.getOrDefault(currCR, Collections.emptySet());
+            for (NodeWithPrev n : starts) {
+                assert(n.getPrev() == null);
+            }
+            q.addAll(starts);
+            while (!q.isEmpty()) {
+                NodeWithPrevAndCost curr = q.poll();
+                if (lcb.equals(curr)) {
+                    List<Node> path = curr.getPrevPath();
+                    allPIPs.addAll(RouterHelper.getPIPsFromNodes(path));
+
+                    Set<NodeWithPrevAndCost> s = startingPoints.get(currCR);
+                    for (Node n : path) {
+                        s.add(new NodeWithPrevAndCost(n));
+                    }
+
+                    continue nextLCB;
+                }
+                for (Node downhill : curr.getAllDownhillNodes()) {
+                    // Stay in this clock region
+                    if (!currCR.equals(downhill.getTile().getClockRegion())) {
+                        continue;
+                    }
+                    IntentCode intentCode = downhill.getIntentCode();
+                    if (intentCode != IntentCode.NODE_PINFEED && !intentCode.isVersalClocking()) {
+                        continue;
+                    }
+                    if (downhill.getWireName().endsWith("_I_CASC_PIN") || downhill.getWireName().endsWith("_CLR_B_PIN")) {
+                        continue;
+                    }
+                    if (!visited.add(downhill)) {
+                        continue;
+                    }
+                    int cost = downhill.getTile().getManhattanDistance(lcbTile);
+                    q.add(new NodeWithPrevAndCost(downhill, curr, cost));
+                }
+            }
+            throw new RuntimeException("ERROR: Couldn't route to distribution line in clock region " + lcb);
+        }
+        clk.getPIPs().addAll(allPIPs);
+    }
+
+    /**
+     * Routes from a GLOBAL_VERTICAL_ROUTE to horizontal distribution lines.
+     * @param clk The clock net to be routed.
+     * @param vroute The node to start the route.
+     * @param clockRegions Target clock regions.
+     * @param down To indicate if it is routing to the group of top clock regions.
+     * @return A list of RouteNodes indicating the reached horizontal distribution lines.
```

Comment:
@WenhaoLin-AMD Can you update this javadoc line please?

Reply:
Have updated this part in https://github.com/Xilinx/RapidWright/pull/1102/commits/823f291b38eb59d5c650690939746d436cc1f290.

I don't know whether we should remove `down`. If we need to merge this method with the one in `UltraScaleClockRouting`, it seems better to keep this variable. Otherwise, we should also remove related javadoc.

Reply:
Thanks. I no longer to plan merging `VersalClockRouting` back into `UltraScaleClockRouting` since they are sufficiently different in being `Node` and `Wire` based respectively.